exchanges/bitmex/bitmex_websocket.go

```diff
@@ -544,122 +542,95 @@ func (b *Bitmex) processOrderbook(data []OrderBookL2, action string, p currency.
-					continue
-				}
-				subscriptions = append(subscriptions, subscription.Subscription{
-					Channel: channels[z] + ":" + pFmt.Format(contracts[y]),
-					Pair:    contracts[y],
-					Asset:   assets[x],
-				})
-			}
-		}
+		assetPairs[a] = p.Format(f)
 	}
-	return subscriptions, nil
-}
 
-// GenerateAuthenticatedSubscriptions Adds authenticated subscriptions to websocket to be handled by ManageSubscriptions()
-func (b *Bitmex) GenerateAuthenticatedSubscriptions() ([]subscription.Subscription, error) {
-	if !b.Websocket.CanUseAuthenticatedEndpoints() {
-		return nil, nil
-	}
+	subs := subscription.List{}
+	for _, baseSub := range b.Features.Subscriptions {
+		if !authed && baseSub.Authenticated {
+			continue
+		}
 
-	pFmt, err := b.GetPairFormat(asset.PerpetualContract, true)
-	if err != nil {
-		return nil, err
-	}
-	contracts, err := b.GetEnabledPairs(asset.PerpetualContract)
-	if err != nil {
-		return nil, err
-	}
-	channels := []string{bitmexWSExecution,
-		bitmexWSPosition,
-	}
-	subscriptions := []subscription.Subscription{
-		{
-			Channel: bitmexWSAffiliate,
-		},
-		{
-			Channel: bitmexWSOrder,
-		},
-		{
-			Channel: bitmexWSMargin,
-		},
-		{
-			Channel: bitmexWSPrivateNotifications,
-		},
-		{
-			Channel: bitmexWSTransact,
-		},
-		{
-			Channel: bitmexWSWallet,
-		},
-	}
-	for i := range channels {
-		for j := range contracts {
-			subscriptions = append(subscriptions, subscription.Subscription{
-				Channel: channels[i] + ":" + pFmt.Format(contracts[j]),
-				Pair:    contracts[j],
-				Asset:   asset.PerpetualContract,
-			})
+		if baseSub.Asset == asset.Empty {
+			subs = append(subs, baseSub.Clone())
+			continue
+		}
+
+		for a, p := range assetPairs {
+			if baseSub.Channel == bitmexWSOrderbookL2 && a == asset.Index {
+				continue // There are no L2 orderbook for index assets
+			}
+			if baseSub.Asset != asset.All && baseSub.Asset != a {
+				continue
+			}
+			s := baseSub.Clone()
+			s.Asset = a
+			s.Pairs = p
+			subs = append(subs, s)
 		}
 	}
-	return subscriptions, nil
+
+	return subs, nil
 }
 
 // Subscribe subscribes to a websocket channel
-func (b *Bitmex) Subscribe(channelsToSubscribe []subscription.Subscription) error {
-	var subscriber WebsocketRequest
-	subscriber.Command = "subscribe"
-	for i := range channelsToSubscribe {
-		subscriber.Arguments = append(subscriber.Arguments,
-			channelsToSubscribe[i].Channel)
+func (b *Bitmex) Subscribe(subs subscription.List) error {
+	req := WebsocketRequest{
+		Command: "subscribe",
 	}
-	err := b.Websocket.Conn.SendJSONMessage(subscriber)
-	if err != nil {
-		return err
+	for _, s := range subs {
+		for _, p := range s.Pairs {
+			cName := channelName(s.Channel)
+			req.Arguments = append(req.Arguments, cName+":"+p.String())
+		}
 	}
-	b.Websocket.AddSuccessfulSubscriptions(channelsToSubscribe...)
-	return nil
+	err := b.Websocket.Conn.SendJSONMessage(req)
+	if err == nil {
+		err = b.Websocket.AddSuccessfulSubscriptions(subs...)
```

Comment:
Oh, this is a good prompt.
In #1516 I introduced an UnsubscribedState, because I think it's wrong to Remove it and not be explicit about the state and Inactive felt very wrong.

So yeah:
* AddSubscription sets State to Subscribing IF not anything else
* AddSuckySubscription sets State to Subscribed
* RemoveSubscription sets state to Unsubscribed

Fixed bebe363f